### PR TITLE
DevOps: Add unit tests for `OrcaBaseParser`

### PR DESCRIPTION
**Pull request by sphuber**
A single test is added for the `OrcaBaseParser`, however, with the new test fixtures it is easy to add additional tests by creating new folders with sample output data in `tests/parsers/fixtures/orca`.

The `OrcaBaseParser` is updated to be compatible with `aiida-core==2.0`.

### CODE REVIEW

**Comment by sphuber** [comment]
@danielhollas

File: aiida_orca/parsers/__init__.py [hunk 1]
```diff
@@ -70,24 +74,18 @@ def _remove_nan(parsed_dictionary: dict) -> dict:
...
-            # relaxation_trajectory = SinglefileData(
-            #     file=os.path.join(out_folder._repository._get_base_folder().abspath, fname_traj)  #pylint: disable=protected-access
-            # )
-            # self.out('relaxation_trajectory', relaxation_trajectory)
 
-        pt = PeriodicTable()  #pylint: disable=invalid-name
+        pt = PeriodicTable()  # pylint: disable=invalid-name
 
-        output_dict['elements'] = [pt.element[Z] for Z in output_dict['atomnos'].tolist()]
+        output_dict['elements'] = [pt.element[Z] for Z in output_dict['atomnos']]
```
Comment (danielhollas) [inline aiida_orca/parsers/__init__.py]: Why was this changed? Would this change work even if we ran a calculation on a single atom?

File: aiida_orca/parsers/__init__.py [hunk 2]
```diff
@@ -24,25 +28,25 @@ def parse(self, **kwargs):
...
 
-        if fname_out not in out_folder._repository.list_object_names():  #pylint: disable=protected-access
+        if fname_out not in out_folder.list_object_names():
             raise OutputParsingError('Orca output file not retrieved')
 
         try:
-            with self.retrieved.open(fname_out) as handler:
-                parsed_obj = ccread(handler.name)
+            with self.retrieved.open(fname_out, 'rb') as handle:
+                with tempfile.NamedTemporaryFile('w+b') as tmpfile:
```
Comment (danielhollas) [inline aiida_orca/parsers/__init__.py:45]: Is the NamedTemporaryFile needed here? In a PR where I implemented the AiiDA 2.0 support, I was able to simply pass the handler directly to `ccread`. 

https://github.com/pzarabadip/aiida-orca/pull/46/files#diff-442a1d8b7896fbb8c11128c00b321b263abe45a0c7134293d4e37f1d22e1a18eR41

File: aiida_orca/parsers/__init__.py [hunk 3]
```diff
@@ -24,25 +28,25 @@ def parse(self, **kwargs):
...
+                    shutil.copyfileobj(handle, tmpfile)
+                    parsed_obj = ccread(tmpfile.name)
                 parsed_dict = parsed_obj.getattributes()
-        except OutputParsingError:  #pylint: disable=bare-except
+        except OutputParsingError:
```
Comment (danielhollas) [inline aiida_orca/parsers/__init__.py]: Could you please change to
```suggestion
        except Exception:
```

The original code was wrong. `OutputParsingError` is an AiiDA type, and is never raised in the `ccread` function.
Parsing here can fail in variety of ways (ccread was extracted from the cclib library) and we need to catch them all. Would be better to be more specific in the future...

File: tests/parsers/test_orca.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    parser = generate_parser(entry_point_parser)
+    results, calcfunction = parser.parse_from_node(node)
+
+    assert calcfunction.is_finished, calcfunction.exception
+    assert calcfunction.is_finished_ok, calcfunction.exit_message
+    assert 'relaxed_structure' in results
+    assert 'output_parameters' in results
+
+    structure_attributes = results['relaxed_structure'].attributes
+    structure_cell = [item for sublist in structure_attributes.pop('cell') for item in sublist]
```
Comment (danielhollas) [inline tests/parsers/test_orca.py]: ORCA doesn't do periodic calculations, so there should be no cell, or a default cell. So I don't think we should check it here.

File: tests/parsers/test_orca.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    data_regression.check({
```
Comment (danielhollas) [inline tests/parsers/test_orca.py:25]: Where do these fixtures come from? What is the intended use case for `data_regression` versus `num_regression`? (I mean I can guess but would like it to be spelled out explicitly)

File: tests/parsers/test_orca/test_orca_default.csv [hunk 6]
```diff
@@ -0,0 +1,10 @@
+,structure_cell
```
Comment (danielhollas) [inline tests/parsers/test_orca/test_orca_default.csv]: I am confused as to where these numbers come from, but I don't think we need this file?

**Review by danielhollas — CHANGES_REQUESTED** [review]
Hi @sphuber,

thanks so so much for this PR, and I apologize that I did not get to it earlier. I just have a couple of questions before merging this.

File: aiida_orca/parsers/__init__.py [hunk 7]
```diff
@@ -24,25 +28,25 @@ def parse(self, **kwargs):
...
-        fname_relaxed = self.node.process_class._RELAX_COORDS_FILE  #pylint: disable=protected-access
-        # fname_traj = self.node.process_class._TRAJECTORY_FILE  #pylint: disable=protected-access
-        # fname_hessian = self.node.process_class._HESSIAN_FILE  #pylint: disable=protected-access
+        process_cls = self.node.process_class
+        fname_out = process_cls._OUTPUT_FILE  # pylint: disable=protected-access
+        fname_relaxed = process_cls._RELAX_COORDS_FILE  # pylint: disable=protected-access
 
-        if fname_out not in out_folder._repository.list_object_names():  #pylint: disable=protected-access
+        if fname_out not in out_folder.list_object_names():
             raise OutputParsingError('Orca output file not retrieved')
```
Comment (danielhollas) [inline aiida_orca/parsers/__init__.py]: I am wondering, why are we raising here, and not returning a non-zero exit code? (as done elsewhere in this file?)

File: aiida_orca/parsers/__init__.py (hunk 1 above)
Comment (sphuber) [reply aiida_orca/parsers/__init__.py]: Think it shouldn't matter because it is a `numpy.ndarray` which is iterable so I don't think conversion to a list is necessary. But I can keep it, it is not a critical change.

File: aiida_orca/parsers/__init__.py (hunk 2 above)
Comment (sphuber) [reply aiida_orca/parsers/__init__.py:45]: But did you test that code? Because if I run the test, it fails:
```
aiida_orca/parsers/cclib/ccio.py:48: in ccopen
    inputfile = logfileparser.openlogfile(source)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

filename = <_io.BufferedReader name='/tmp/tmpxsaxb48r/test_repo/repository/node/2b/77/4f79-56b5-46a1-b222-7882881e6b1c/path/aiida.out'>, object = None

    def openlogfile(filename, object=None):
        """Return a file object given a filename or if object specified decompresses it
        if needed and wrap it up.
    
        Given the filename or file object of a log file or a gzipped, zipped, or bzipped
        log file, this function returns a file-like object.
    
        Given a list of filenames, this function returns a FileInput object,
        which can be used for seamless iteration without concatenation.
        """
    
>       fileobject = FileWrapper(io.open(filename, 'r', errors='ignore'))
E       TypeError: expected str, bytes or os.PathLike object, not _io.BufferedReader
```
It is calling `io.open` which doesn't support file handles.

File: aiida_orca/parsers/__init__.py (hunk 7 above)
Comment (sphuber) [reply aiida_orca/parsers/__init__.py]: Fully agree, this was the original code. I will add an exit code for it

File: aiida_orca/parsers/__init__.py (hunk 3 above)
Comment (sphuber) [reply aiida_orca/parsers/__init__.py]: Done

File: tests/parsers/test_orca.py (hunk 5 above)
Comment (sphuber) [reply tests/parsers/test_orca.py:25]: They come from the `pytest-regressions` package (https://pypi.org/project/pytest-regressions/). The `data_regression` is used for generic dictionary data structures whereas `num_regression` is used for numpy arrays: https://pytest-regressions.readthedocs.io/en/latest/api.html#num-regression

File: tests/parsers/test_orca/test_orca_default.csv (hunk 6 above)
Comment (sphuber) [reply tests/parsers/test_orca/test_orca_default.csv]: This is the file that is automatically generated by the `data_regression` fixture. The first time the test was run, this reference file was automatically created. Now each time the test is run, the test result is automatically compared to this reference. If the code changes such that the result is changed, the test fails. If this is for expected reasons, the reference file can be automatically regenerated by passing `--force-regen` to the `pytest` call. This is very convenient to write tests against arbitrary data structures.

File: tests/parsers/test_orca.py (hunk 4 above)
Comment (sphuber) [reply tests/parsers/test_orca.py]: Removed

Comment (sphuber) [reply tests/parsers/test_orca.py]: Note that apparently the parsed output data has a structure that actually _does_ define a cell. If this is non-sensical, maybe the parser should be updated to not create this cell in the output structure?

**Comment by sphuber** [comment]
@danielhollas thanks for the review. I have addressed your comments.

File: aiida_orca/parsers/__init__.py (hunk 2 above)
Comment (danielhollas) [reply aiida_orca/parsers/__init__.py:45]: Ah, in that PR I got rid of the `logfileparser.openlogfile()` call altogether as it is not needed, and that's I think how I made it work. Never mind then, I'll deal with this later after these changes are merged.

File: tests/parsers/test_orca.py (hunk 5 above)
Comment (danielhollas) [reply tests/parsers/test_orca.py:25]: Thank you for the explanation, that is very useful.

File: aiida_orca/parsers/__init__.py (hunk 7 above)
Comment (danielhollas) [reply aiida_orca/parsers/__init__.py]: Perfect, thank you!

File: tests/parsers/test_orca.py (hunk 4 above)
Comment (danielhollas) [reply tests/parsers/test_orca.py]: @sphuber It looks like removing this line broke the test for AiiDA 1.6. I think you still need to pop the `cell` attribute (but not test it), since before AiiDA 2.0, the `StructureData` nodes required non-zero cell even for non-periodic structures. Which is why I think you might have seen it there. 

The current parser uses pymatgen to read an XYZ file from ORCA, which definitely doesn't define any cell. In the other PR I am getting rid of pymatgen anyway so we don't need to pursue this here.

File: tests/parsers/test_orca/test_orca_default.csv (hunk 6 above)
Comment (danielhollas) [reply tests/parsers/test_orca/test_orca_default.csv]: Thank you for the explanation. Since we (`aiida-orca` devs) are new into these pytest fixtures / plugins, it would be very useful to include this info somewhere, either as a comment in the conftest file, or in a dev part of the main README perhaps?

**Comment by danielhollas** [comment]
@sphuber thanks so much! I think one more fix is needed to make the tests pass for AiiDA 1.x, but otherwise looks good!

File: tests/parsers/test_orca/test_orca_default.csv (hunk 6 above)
Comment (sphuber) [reply tests/parsers/test_orca/test_orca_default.csv]: Will add it to the `README.md` for now. If it becomes too big at some point you can always move it out somewhere else

Comment (sphuber) [reply tests/parsers/test_orca/test_orca_default.csv]: Added

File: tests/parsers/test_orca.py (hunk 4 above)
Comment (sphuber) [reply tests/parsers/test_orca.py]: Popped the `cell` and updated the reference file

**Comment by sphuber** [comment]
Thanks @danielhollas should be good to go now.

**Review by danielhollas — APPROVED** [review]
Great, thanks @sphuber!
